<launch>

  <!-- Arguments -->

  <!-- Image arguments -->
  <arg name = "resolution" default = "full"/>
  <arg name = "stereo" default = "stereo_down"/>
  <arg name = "camera" default = "/stereo_down"/>
  <arg name = "enable_decimate_x1" default = "false"/>
  <arg name = "enable_decimate_x2" default = "true"/>
  <arg name = "enable_decimate_x4" default = "false"/>
  <arg name = "images_approximate_sync" default = "true"/>

  <!-- Odometer arguments -->
  <arg name = "viso2_processor" default = "false"/>
  <arg name = "bucketing" default = "false"/>
  <arg name = "combination" value = "2"/> <!-- 0: SIFT_SIFT, 1: SURF_SIFT, 2: SURF_BRISK -->
  <arg name = "epipolar_constrain" value = "1"/>
  <arg name = "contrast_threshold" value = "0.04"/> <!-- Only when SIFT is used as feature detector -->
  <arg name = "min_hessian" value = "1500"/> <!-- Only when SURF is used as feature detector. 
                                                  Recommended value for 1: 1750.
                                                  Recommended value for 2: 1500. -->

  <!-- Frame arguments -->
  <arg name = "base_link" default = "turbot/base_link"/>
  <arg name = "sensor" default = "turbot/stereo_down/left_optical"/>
  <!-- <arg name = "base_link" default = "sparus2"/>
  <arg name = "sensor" default = "left_optical"/> -->

  <!-- Parameters -->
  <param name="/use_sim_time" value = "true"/>

  <!-- Bagfiles -->

  <!-- Portals Vells -->
  <arg name = "bagfile" default = "/home/uib/bagfiles/portals/2021_02_25/14_41_06/images_2021-02-25-14-41-06_0.bag
                                    /home/uib/bagfiles/portals/2021_02_25/14_41_06/images_2021-02-25-14-47-26_1.bag
                                    /home/uib/bagfiles/portals/2021_02_25/14_41_06/images_2021-02-25-14-53-45_2.bag
                                    /home/uib/bagfiles/portals/2021_02_25/14_41_06/images_2021-02-25-15-00-04_3.bag
                                    /home/uib/bagfiles/portals/2021_02_25/14_41_06/images_2021-02-25-15-06-23_4.bag
                                    /home/uib/bagfiles/portals/2021_02_25/14_41_06/images_2021-02-25-15-12-43_5.bag"/>
                                    <!-- /home/uib/bagfiles/portals/2021_02_25/14_41_06/topics_2021-02-25-14-41-52_0.bag"/> -->

  <!-- Valldemossa -->
  <!-- <arg name="bagfile" default="/home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0000.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0001.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0002.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0003.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0004.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0005.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0006.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0007.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0008.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0009.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0010.bag
                                  /home/uib/bagfiles/valldemossa/2016_09_21_valldemossa/2016_09_21_valldemossa/160921_1659/20160921_165907_0011.bag"/>  -->

  <!-- Play the bagfile -->
  <!-- <node pkg = "rosbag" type = "play" name = "rosbag" args = "-r 0.25 clock $(arg bagfile)" /> -->
  <node pkg = "rosbag" type = "play" name = "rosbag" args = "-r 1 --clock $(arg bagfile)" />

  <!-- Run the stereo image proc -->
  <group>
    <include file="/home/uib/catkin_ws/src/image_pipeline/stereo_image_proc/launch/stereo_proc.launch">
    <arg name = "stereo" value = "$(arg stereo)" />
    <arg name = "resolution" value = "$(arg resolution)" />
    <arg name = "enable_decimate_x1" value = "$(arg enable_decimate_x1)" />
    <arg name = "enable_decimate_x2" value = "$(arg enable_decimate_x2)" />
    <arg name = "enable_decimate_x4" value = "$(arg enable_decimate_x4)" />
    </include>
  </group>

  <!-- Static TF's -->
  <node pkg="tf" type="static_transform_publisher" name="turbot_to_stereo_stick" args="0.4 0.0 0.8 0.0 0.0 0.0 /turbot/base_link /turbot/stereo_stick 30" />
  <node pkg="tf" type="static_transform_publisher" name="stereo_stick_to_stereo_base" args="0.0 0.0 0.0 1.57 0.0 1.57 /turbot/stereo_stick /turbot/stereo_down/base 30" />
  <node pkg="tf" type="static_transform_publisher" name="stereo_base_to_stereo" args="0.0 0.0 0.0 0.0 0.0 -1.57 /turbot/stereo_down/base /turbot/stereo_down 30" />
  <node pkg="tf" type="static_transform_publisher" name="stereo_to_left" args="-0.06 0.0 0.0 0.0 0.0 0.0 /turbot/stereo_down /turbot/stereo_down/left_optical 30" />


  <!-- Viso2 -->
  <node pkg = "viso2_ros" type = "stereo_odometer" name = "stereo_odometer" output = "screen" if = "$(eval enable_decimate_x1 == true)">
    <remap from = "stereo" to = "$(arg camera)"/>
    <remap from = "image" to = "image_rect"/>
    <param name = "queue_size" value = "5"/>
    <param name = "approximate_sync" value = "$(arg images_approximate_sync)"/>
    <param name = "Viso2_processor" value = "$(arg viso2_processor)"/>
    <param name = "Bucketing" value = "$(arg bucketing)"/>
    <param name = "Combination" value = "$(arg combination)"/>
    <param name = "Epipolar_constrain" value = "$(arg epipolar_constrain)"/>
    <param name = "Contrast_threshold" value = "$(arg contrast_threshold)"/>
    <param name = "Min_hessian" value = "$(arg min_hessian)"/>
    <param name = "base_link_frame_id" value = "$(arg base_link)"/>
    <param name = "sensor_frame_id" value = "$(arg sensor)"/>
  </node>

  <node pkg = "viso2_ros" type = "stereo_odometer" name = "stereo_odometer" output = "screen" if = "$(eval enable_decimate_x2 == true)">
    <remap from = "stereo" to = "$(arg camera)"/>
    <remap from = "/stereo_down/left/image_rect" to = "/stereo_down/scaled_x2/left/image_rect"/>
    <remap from = "/stereo_down/right/image_rect" to = "/stereo_down/scaled_x2/right/image_rect"/> 
    <remap from = "image" to = "image_rect"/>
    <param name = "queue_size" value = "5"/>
    <param name = "approximate_sync" value = "$(arg images_approximate_sync)"/>
    <param name = "Viso2_processor" value = "$(arg viso2_processor)"/>
    <param name = "Bucketing" value = "$(arg bucketing)"/>
    <param name = "Combination" value = "$(arg combination)"/>
    <param name = "Epipolar_constrain" value = "$(arg epipolar_constrain)"/>
    <param name = "Contrast_threshold" value = "$(arg contrast_threshold)"/>
    <param name = "Min_hessian" value = "$(arg min_hessian)"/>
    <param name = "base_link_frame_id" value = "$(arg base_link)"/>
    <param name = "sensor_frame_id" value = "$(arg sensor)"/>
  </node>

  <node pkg = "viso2_ros" type = "stereo_odometer" name = "stereo_odometer" output = "screen" if = "$(eval enable_decimate_x4 == true)">
    <remap from = "stereo" to = "$(arg camera)"/>
    <remap from = "/stereo_down/left/image_rect" to = "/stereo_down/scaled_x4/left/image_rect"/>
    <remap from = "/stereo_down/right/image_rect" to = "/stereo_down/scaled_x4/right/image_rect"/> 
    <remap from = "image" to = "image_rect"/>
    <param name = "queue_size" value = "5"/>
    <param name = "approximate_sync" value = "$(arg images_approximate_sync)"/>
    <param name = "Viso2_processor" value = "$(arg viso2_processor)"/>
    <param name = "Bucketing" value = "$(arg bucketing)"/>
    <param name = "Combination" value = "$(arg combination)"/>
    <param name = "Epipolar_constrain" value = "$(arg epipolar_constrain)"/>
    <param name = "Contrast_threshold" value = "$(arg contrast_threshold)"/>
    <param name = "Min_hessian" value = "$(arg min_hessian)"/>
    <param name = "base_link_frame_id" value = "$(arg base_link)"/>
    <param name = "sensor_frame_id" value = "$(arg sensor)"/>
  </node>


</launch>
